runtime/ftplugin/javascript.vim

```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
 
 setlocal commentstring=//%s
 
-" Change the :browse e filter to primarily show Java-related files.
+" Change the :browse e filter to primarily show JavaScript-related files.
 if has("gui_win32")
-    let  b:browsefilter="Javascript Files (*.js)\t*.js\n" .
-		\	"All Files (*.*)\t*.*\n"
+    let  b:browsefilter="Javascript Files (*.js, *.jsx, *.es, *.es6, *.cjs, *.mjs, *.jsm, *.vue, *.json)\t*.js;*.jsx;*.es;*.es6;*.cjs;*.mjs;*.jsm;*.vue;*.json\n" .
+                \ "All Files (*.*)\t*.*\n"
 endif
-       
-let b:undo_ftplugin = "setl fo< ofu< com< cms<" 
+
+" The following suffixes should be implied when resolving filenames
+setlocal suffixesadd+=.js,.jsx,.es,.es6,.cjs,.mjs,.jsm,.vue,.json
+
+" The following suffixes should have low priority
+"   .snap    jest snapshot
+setlocal suffixes+=.snap
+
+" Prepend node_modules/.bin to $PATH if applicable
+" Allows calling executables installed locally via npm/yarn
+let s:bin_dir = finddir('node_modules/.bin', '.;')->fnamemodify(':p')
+if len(s:bin_dir) && $PATH !~ s:bin_dir
+    let $PATH = s:bin_dir .. ':' .. $PATH
+endif
+unlet s:bin_dir
+
```

Comment:
Without this it is very hard to use linters or formatters as it is customary to include them as project dependencies.

Maybe it would be possible to perform path resolution at the formatter/compiler level?

Reply:
I know what's the purpose of `$PATH` and adding specific directories to `$PATH` fits that purpose entirely. Again, without this locally-installed linters and formatters can't be called.

Please suggest a better approach that, ideally, doesn't require writing a hundred lines of documentation that the user will never read.

> A quick search suggests that this would be the only place in the entire runtime distribution that $PATH was modified.

JavaScript is one of the rare ecosystem where that might be useful.

Reply:
> By "system PATH environment variable" I meant it should be properly configured in the environment from which Vim is started. This is the better approach.

A better approach that prevents use of linters and formatters that are installed locally.

> Even in an ideal world, the PATH component you're prepending, being relative, is really only useful when Vim's working directory is fixed.

It is absolute.

> I suppose if you want to go all in on Node.js you could have the compilers use npx but that wouldn't really be my preference. It's probably a rather slippery slope.

I don't want to go that way, hence this.

> JavaScript is not special. ;)

The language isn't. The ecosystem and the established practices are.

Reply:
@dkearns, ESLint and Standard are currently the most widely used linters in JavaScript land and beyond (ESLint is now the official TypeScript linter). You are right that people often use plugins to interact with those external tools but that has more to do with the state of abandonment of the JavaScript runtime than with technical limitations.

Now that I have added proper "compilers" for ESLint and Standard to `$VIMRUNTIME/compiler/` all that is left is to make them usable in the most seamless way possible.

But we need a quick reality check, first: those tools can certainly be installed globally but they are more often than not installed *locally*, as part of project dependencies. This means that:

- if the tool is installed globally it is likely to be somewhere in `$PATH`,
- if the tool is installed locally it is likely to not be in `$PATH`.

The second case is extremely easy to handle with minimal overhead and that's what I did here:

1. find the absolute path to the parent `node_modules/.bin`,
2. check if it is already in `$PATH`,
3. if not, prepend it to `$PATH`.

The idea is to make sure locally-installed tools are as accessible and usable as if they were installed globally. It is totally transparent and requires no user action.

With that setup and the `:compiler` commands in the ftplugin, here is what the user would have to do before being able to use of the built-in `eslint` compiler:

1. edit a JavaScript file.

By contrast, here is what the user would have to do before being able to use of the built-in `eslint` compiler:

1. do `$ PATH=<absolute path to node_modules/.bin/>:$PATH` either before starting Vim or from Vim itself,
2. do `:compiler eslint` in every relevant buffer, or do `:compiler! eslint` for *every* buffer (with caveat).

Yes, that's not a really a common strategy in `$VIMRUNTIME/` but that strategy is, in my opinion, a significant upgrade to the existing status quo.